rclcpp/include/rclcpp/publisher_options.hpp

```diff
@@ -102,7 +104,12 @@ struct PublisherOptionsWithAllocator : public PublisherOptionsBase
   get_allocator() const
   {
     if (!this->allocator) {
-      return std::make_shared<Allocator>();
+      if constexpr (std::is_same_v<Allocator, std::allocator<void>>) {
+        auto g_instance = std::make_shared<std::allocator<void>>();
+        return g_instance;
+      } else {
+        throw std::runtime_error("allocator is nullptr");
```

Comment:
Why can we not make an allocator if it isn't std::allocator?

Reply:
With stateless allocators, a new instance is the same than a copy of an old instance.
A custom allocator might be stateful, and thus a new instance and a copy of an old instance might not be the same.

I wondered about the possibility of `get_allocator()` being called multiple times, and retrieving a difference instance each time, which would be unexpected.

I didn't like the `static` variable alternative, because it might use resources innecesarly if `options.allocator != nullptr` and the allocator is stateful (and statics are not guaranteed to be initialized lazely).

Reply:
I think Ivan's right. I don't see a solution for stateful custom allocators other than throwing like this. Attempting lazy initialization has its own set of issues, as discussed above, and won't do for allocators that are not default-constructible (AFAIK no allocator is required to).

I will say though that using `if constexpr` prevents this patch from being backported to Foxy.

Reply:
Well, I think it comes down to whether or not we return multiple copies. I agree that's a problem, but if we return only one new instance and reuse it from then on, then I don't think there's a problem in making one on demand.

Reply:
>  I don't think there's a problem in making one on demand. 

Assuming you can i.e. that `Allocator` is default constructible, which need not be.

That's why I'd say we keep this patch simple, throwing or failing compilation for allocators that are not default constructible like Ivan did, and returning copies for the rest. That we can backport. For H-Turtle we should revamp all of this.

Reply:
> and statics are not guaranteed to be initialized lazely

Sorry, this is actually wrong, statics are guaranteed to be lazely initialized (see https://en.cppreference.com/w/cpp/language/storage_duration#Static_local_variables).

It would still be different of what `std::vector` do, which is to use a new `Allocator()` for each `std::vector<Allocator>` that didn't explicitly provide one, instead of sharing the same static one.
But maybe that's better than throwing (?).

We could also use an extra private field for the extra storage (see first part of [this comment](https://github.com/ros2/rclcpp/pull/1647#discussion_r625739061), it would only introduce an ABI break (so we cannot backport to Foxy, but I think we can still do that in galactic).

Reply:
> We could also use an extra private field for the extra storage (see first part of this comment, it would only introduce an ABI break (so we cannot backport to Foxy, but I think we can still do that in galactic).

That'd be my vote... Another reason we should really do the pimpl everywhere so we have options in the future.

Reply:
Fixed in https://github.com/ros2/rclcpp/pull/1647/commits/fe13993311aa5f0ce43b60039503a019b84353f4.